More immune regex

allejok96 opened the pull request:
Fixes these bugs and adds ability to debug a single file

```cpp
} // unexpected closing brace causes IndexError

namespace ShortNamespace {
#ifdef // this macro should throw an error but it will never be parsed
// this matches --> }  even if it is a comment
}

{ // unclosed brace does not throw error
```

JohannesLorenz commented:
Thanks a lot.

In theory, this looks good, but it does not pass `verify`'s tests. If you just run `tests/scripted/verify`, currently it passes. Change the `if False:` to `if True:`, and ~3 more tests will be activated that fail. Can you please try them out?

allejok96 commented:
Looking at the verify script I understand a few more things.

When I made the regex I changed `#include inside namespace` to `#include after beginning of C++ code`. I thought it was a style guide but now I realize it prevents a very real problem. Should I change it back or is it good to enforce having all includes at the top?

I also found another small bug in my regex ;) Other than that, it's just a matter of updating the error messages.

But the forgive my saying the new verify script feels a bit over-complicated in its approach. Updating the error messages now requires re-running the check for each message and adding a new error is several lines of code.

I'd suggest having "virtual test environment" in `tests/src` with files like `BadNamespace.h` and `GoodNamespace.h`. If possible only calling `check-namespace` and `check-strings` once and then looking at the output. These files would also serve as examples on how the could should and shouldn't look.

JohannesLorenz commented:
> When I made the regex I changed `#include inside namespace` to `#include after beginning of C++ code`. I thought it was a style guide but now I realize it prevents a very real problem. Should I change it back or is it good to enforce having all includes at the top?

Which very real problem do you mean? Also, are there cases where our code-base uses include after C++ code intentionally (i.e. your new script finds them)?

> But the forgive my saying the new verify script feels a bit over-complicated in its approach. Updating the error messages now requires re-running the check for each message and adding a new error is several lines of code.

I don't understand 100% of what you're saying. The new verify script runs multiple code examples, so it indeed applies all checks multiple times (each check once on each example code), however, you can simply grep with `test.expect()` for those errors that you want to test and do not need to worry about others. So adding a new error should mean adding a new minimal example and callign test.expect for this one new error?

allejok96 commented:
> Which very real problem do you mean?

Including code inside a namespace may break things I guess... idk

> Also, are there cases where our code-base uses include after C++ code intentionally

Four files does this, but I can't find a good reason. One even having `#include` inside an `extern "C"`, stackoverflow says that's a bad idea.

> those errors that you want to test and do not need to worry about others

What's the worry? Why not just have one file full of errors and just check all at once (when possible to combine errors of course)? And why hard-code code examples and have python create them each run, instead of just storing them as a file? Only reason I see for that is that `git init` requires a temporary directory. But only reason for that is to use `git ls-tree` to get a list of files to test. IMO we could skip that.

allejok96 commented:
Mockup of a simple `verify` like the previous one

`check-strings`

```py
if sys.argv:
    files = sys.argv
else:
    files = subprocess.run("git ls-tree")

# Continue with checks...
```

`verify`

```py
exit_status = 0

# Somehow get list of files
files = Path("test").glob("*")

expectation = """
badfile: Missing comment
15 errors
"""

result = subprocess.run(["check-strings"] + files)

for line in expectations:
    if line not in result:
        print("Error")
        exit_status = 1

exit(exit_status)

JohannesLorenz commented:
> > Also, are there cases where our code-base uses include after C++ code intentionally
> 
> Four files does this, but I can't find a good reason. One even having `#include` inside an `extern "C"`, stackoverflow says that's a bad idea.

In this case, I'd leave it as it is now. Don't change it back.

> > those errors that you want to test and do not need to worry about others
> 
> What's the worry? Why not just have one file full of errors and just check all at once (when possible to combine errors of course)? And why hard-code code examples and have python create them each run, instead of just storing them as a file? Only reason I see for that is that `git init` requires a temporary directory. But only reason for that is to use `git ls-tree` to get a list of files to test. IMO we could skip that.

If you have only one test file (instead of having one MWE per error message), the problem can arise that two errors exclude each other, or one error triggers another "false positive" - so if we do it like in your proposal, we might be forced to use multiple tests again. Plus, I find minimal examples more reable.

On top of that, having the expectations right under the MWE in the same file prevents us from switching between example code and expectations.

Anyways, this is all opinion based. I like my current solution best, but won't keep you from rewriting it.

allejok96 commented:
> in the same file prevents us from switching between example code and expectations

Good point. Anyway your current solution works, so if you're satisfied, let's go with it.

>> Four files does this, but I can't find a good reason. One even having #include inside an extern "C", stackoverflow says that's a bad idea.
>
> In this case, I'd leave it as it is now. Don't change it back.

What? Should we fix the `extern` thing or remove the `#include` after code restriction?

JohannesLorenz commented:
> What? Should we fix the `extern` thing or remove the `#include` after code restriction?

Fix the extern thing. And the many other issues that arise with this PR on the current code.

And if we accept the current design of `verify`, another action point is to get the currently outcommented tests running.

allejok96 commented:
Sorry it may look like I re-wrote the entire `check-namespace` again :crying_cat_face: but the most of the diff is due to switching to a `NamedTuple`. The diff got so messy anyway so instead of trying to be graceful I took the opportunity to rename and rearrange the code so it got more readable.

Anyway, the actual improvements are:

- Breaks on syntax error when parsing file (no additional false alarms)
- `#include` not allowed inside `{}` with the exception of `extern "C"`
- Checks malformed comments of header guards (but like before it doesn't require comments for them)
- Updated `verify` script with these checks

Regarding `verify` I added both some good and bad code in the examples to make sure it errors on the bad code only.

allejok96 commented:
Btw there are 57 files with errors now. I have a diff to fix these if you want, but I'm not including it in this PR.

JohannesLorenz commented:
Perfect. Verification looks good. I'll just merge this one and then you continue with the 57 files in the main PR?

allejok96 commented:
Alright
